ip_challenge/Ip_Address_challenge.py

```diff
@@ -1,165 +1,165 @@
+            if user_code and self.validate_function(user_code, test_cases):
+                self.node_levels[node_pos] += 1
+                if self.node_levels[node_pos] >= len(self.node_problems[node_pos]):
+                    return True, f"All problems solved at node: {node_pos}"
+                else:
+                    return False, f"Level {self.node_levels[node_pos]} completed at node: {node_pos}"
+            else:
+                return False, "Incorrect answer or function did not pass the test cases. Try again."
+        return False, None
+
+    def run(self):
+        running = True
+        while running:
+            for event in pygame.event.get():
+                if event.type == pygame.QUIT:
+                    running = False
+                elif event.type == pygame.MOUSEBUTTONDOWN and event.button == 1:
+                    pos = pygame.mouse.get_pos()
+                    if self.show_summary:
+                        if self.button_rect.collidepoint(pos):
+                            self.show_summary = False
+                            self.present_challenge((700, 100))
+                    else:
+                        for rect, node_pos in self.nodes:
+                            if rect.collidepoint(pos):
+                                success, message = self.present_challenge(node_pos)
+                                print(message)
+
+            self.screen.fill((255, 255, 255))
+
+            if self.show_summary:
+                self.draw_summary()
+            else:
+                self.draw_challenge()
+
+            pygame.display.flip()
+
+        pygame.quit()
+
+    def draw_summary(self):
+        title = self.font.render("Welcome to the IP Address Challenge:", True, (0, 0, 0))
+        description = self.font.render("Press the button below to start the challenge.", True, (0, 0, 0))
+
+        # Center the title and description
+        title_rect = title.get_rect(center=(self.screen_width // 2, self.screen_height // 2 - 100))
+        description_rect = description.get_rect(center=(self.screen_width // 2, self.screen_height // 2))
+
+        self.screen.blit(title, title_rect)
+        self.screen.blit(description, description_rect)
+
+        # Center the button
+        self.button_rect = pygame.Rect(0, 0, 200, 50)
+        self.button_rect.center = (self.screen_width // 2, self.screen_height // 2 + 100)
+
+        pygame.draw.rect(self.screen, (0, 255, 0), self.button_rect)
+        button_text = self.button_font.render("Start Challenge", True, (255, 255, 255))
+        button_text_rect = button_text.get_rect(center=self.button_rect.center)
+        self.screen.blit(button_text, button_text_rect)
+
+    def draw_challenge(self):
+        for rect, node_pos in self.nodes:
+            pygame.draw.rect(self.screen, (0, 0, 255), rect)
+            text = self.font.render(f"Node {node_pos}", True, (255, 255, 255))
+            self.screen.blit(text, rect.topleft)
+
```

Comment:
add this function as well

```
def is_completed(self):
     return self.completed
```

this function just alerts the code outside this class if this challemge has been successfullly completed.